### PR TITLE
Introduce `yield_groups()`

### PR DESCRIPTION
**Pull request by lippserd**
This PR introduces `yield_groups()` to yield sets of items from a sorted iterator grouped by a specific criterion gathered from a callback:

```php
namespace ipl\Html\Example;

use Generator;
use Iterator;

use function ipl\Stdlib\yield_groups;

class YieldGroupsExample
{
    /**
     * @param Iterator<int, object{category: int, id: string, data: string}> $data
     *
     * @return Generator<int, array<int, object{category: int, id: string, data: string}>>
     */
    public function groupByCategory(Iterator $data): Generator
    {
        return yield_groups($data, function ($v): int {
            return $v->category;
        });
    }

    /**
     * @param Iterator<int, object{category: int, id: string, data: string}> $data
     *
     * @return Generator<int, array<int, string>>
     */
    public function groupByCategoryWithDataAsValue(Iterator $data): Generator
    {
        return yield_groups($data, function ($v): array {
            return [$v->category, $v->data];
        });
    }

    /**
     * @param Iterator<int, object{category: int, id: string, data: string}> $data
     *
     * @return Generator<int, array<string, string>>
     */
    public function groupByCategoryWithDataAsValueAndIdAsKey(Iterator $data): Generator
    {
        return yield_groups($data, function ($v): array {
            return [$v->category, $v->data, $v->id];
        });
    }
}

### CODE REVIEW

File: src/functions.php [hunk 1]
```diff
@@ -69,3 +71,41 @@ function iterable_key_first($iterable)
...
+ * but can also return value and key in addition.
+ *
+ * @param Iterator $iterator
+ * @param callable(mixed $value, mixed $key): array{0: mixed, 1?: mixed, 2?: mixed} $groupBy
+ *
+ * @return Generator
+ */
+function yield_groups(Iterator $iterator, callable $groupBy): Generator
+{
+    if (! $iterator->valid()) {
```
Comment (yhabteab) [inline src/functions.php:88]: This requirement (the passed `$iterator` being an instance of `Iterator`) does not allow to easily pass objects that implement the `IteratorAggregate` interface, as our IPL orm class `Query` does. You now have to explicitly call the `getIterator()` method. Can't we make this a little less strict?

File: tests/FunctionsTest.php [hunk 2]
```diff
@@ -98,4 +130,118 @@ public function testIterableKeyFirstReturnsNullIfIterableIsGeneratorAndIsEmpty()
             yield;
         })));
     }
+
+    public function testYieldGroupsWithEmptyIterator()
+    {
+        $this->assertEquals([], iterator_to_array(Stdlib\yield_groups(new EmptyIterator(), function () {
+        })));
```
Comment (yhabteab) [inline tests/FunctionsTest.php:137]:
```suggestion
        $this->assertEquals([], iterator_to_array(Stdlib\yield_groups(new EmptyIterator(), function () {})));
```

File: tests/FunctionsTest.php [hunk 3]
```diff
@@ -98,4 +130,118 @@ public function testIterableKeyFirstReturnsNullIfIterableIsGeneratorAndIsEmpty()
...
+    }
+
+    public function testYieldGroupsCallbackArguments()
+    {
+        Stdlib\yield_groups(new ArrayIterator(static::YIELD_GROUPS_DATA), function (array $v, string $k): int {
+            $this->assertEquals(static::YIELD_GROUPS_DATA[$k], $v);
+
+            return $v['group'];
+        });
```
Comment (yhabteab) [inline tests/FunctionsTest.php:146]: Forgot to assert that :)

**Review by yhabteab — CHANGES_REQUESTED** [review]

File: src/functions.php (hunk 1 above)
Comment (lippserd) [reply src/functions.php:88]: Sure, `Traversable` + `IteratorIterator`.

File: tests/FunctionsTest.php (hunk 2 above)
Comment (lippserd) [reply tests/FunctionsTest.php:137]: I would rather go with `$this->assertEquals([], iterator_to_array(Stdlib\yield_groups(new EmptyIterator(), function () {})));`

File: tests/FunctionsTest.php (hunk 3 above)
Comment (lippserd) [reply tests/FunctionsTest.php:146]: No. This test is about the callback arguments. Not the return.

File: tests/FunctionsTest.php (hunk 2 above)
Comment (lippserd) [reply tests/FunctionsTest.php]: Ok, that is not allowed 😆

File: tests/FunctionsTest.php (hunk 3 above)
Comment (yhabteab) [reply tests/FunctionsTest.php]: The assertion inside the callback is never run and I don't know how the GitHub actions even succeeded.

Comment (lippserd) [reply tests/FunctionsTest.php]: That is so true.

Comment (yhabteab) [reply tests/FunctionsTest.php]:
```bash
This test did not perform any assertions

/usr/local/src/ipl-stdlib/tests/FunctionsTest.php:140
```

File: tests/FunctionsTest.php (hunk 2 above)
Comment (lippserd) [reply tests/FunctionsTest.php:137]: Ah. I think you've requested to change it to be a oneliner. Our style checks fail then.

**Review by yhabteab — DISMISSED** [review]
Tested with: https://github.com/Icinga/icingaweb2-module-x509/pull/185/files#diff-e248f67b3802079b7e7cfc304f1873fba6c74e354a37c22e6bc56443a15f813dR505-R512

**Comment by nilmerg** [comment]
I've just fixed the phpstan complaints: https://github.com/Icinga/ipl-stdlib/actions/runs/5975237958

**Review by nilmerg — APPROVED** [review]

**Review by yhabteab — APPROVED** [review]

**Pull request merged**